File: medium/longestSubstring.c
// this is brute force method

#include<stdio.h>
#include<stdlib.h>
#include<string.h>
char* appendCharToCharArray(char* array, char *a){
    int newSize = strlen(array)  + strlen(array) + 1; 
    char *newBuffer = (char *)malloc(newSize);

    // do the copy and concat
    strcpy(newBuffer,array);
    strcat(newBuffer,a); 
    
    array = newBuffer;
    
    
    return array;
}
int checkInString(char *input,char c){
    int len = strlen(input);
    for (int i = 0;i<len;i++){
        if(input[i]==c){
            //printf("is in\n");
            return 1;
        }else{
            //printf("no, it is not in string\n");
            continue;
        }
    }
    return 0;
}
int lengthOfLongestSubstringByBruteForce(char *input) {

    
    int len = strlen(input);
    
    
    

    char str[1000000];
    
    char longest[10000];
    
    
    //input = appendCharToCharArray(input,a);

    memcpy(str,&input[0],len);
    str[len]='\0';
    
    memcpy(longest,&input[0],1);
    longest[1]='\0';
    
    //printf("str = %s\n",str);
    //printf("longest =%s\n",longest);
    //printf("str[1] = %c\n",str[1]);
    
    for (int head = 0;head<len;head++){
        // initial tmp = input[head]
        // it is mean that head flag move to next location and try to find
        // longest substring
        char tmp[10000];
        memcpy(tmp,&input[head],1);
        tmp[1]='\0';

        //printf("head %d: \n ",head);

        for (int i =head+1;i<len;i++){
            // check str[i] is in tmp? if yes, it is repeat. 
            // if no, it is longer
            if(checkInString(tmp,str[i])==0){

                //printf("str[%d] = %c is not in tmp : %s\n",i,str[i],tmp);
                memcpy(tmp,&str[head],i+1-head);
                tmp[i+1-head]='\0';
                //printf("tmp = %c  %c  %s\n",tmp[1],tmp[2],tmp);
                
                
            }else{ // equal
                //printf("str[%d]=%c is in tmp : %s\n",i,str[i],tmp);
                break;
                
            }
            
        }
        // check current substring with current head is longer than old longest?
        if(strlen(tmp)>strlen(longest)){
            memcpy(longest,tmp,strlen(tmp)+1);
        }
        
    }
    

    printf("longest = %s ans = %lu \n",longest,strlen(longest));
    
    return strlen(longest);
}


int main(void){

    //printf("sss\n");
    char* input ="wobgrovw";
    int ans = lengthOfLongestSubstring(input);
    printf("ans = %d", ans);
}

File: medium/longestSubstring.py
class Solution(object):
    def lengthOfLongestSubstring(self, s):
        """
        :type s: str
        :rtype: int
        """
        ## Runtime: 6188 ms, faster than 5.01% of Python online submissions for Longest Substring Without Repeating Characters.
        ## Memory Usage: 13.8 MB, less than 6.25% of Python online submissions for Longest Substring Without Repeating Characters.
        ## brute method
        '''
        if not s:
            return 0
        longest = s[0]
        
        for i in range(0,len(s)):
            current = ''
            for j in range(i,len(s)):
                if s[j] not in current:
                    current = current + s[j] 
                else:
                    break
            print('i = {} current = {}'.format(i, current))
            if len(current) > len(longest):
                longest = current 
            
        return len(longest)
        '''

        ## Runtime: 564 ms, faster than 11.20% of Python online submissions for Longest Substring Without Repeating Characters.
        ##Memory Usage: 12.1 MB, less than 59.38% of Python online submissions for Longest Substring Without Repeating Characters.

        if not s:
            return 0
        longest = s[0]
        i = 0
        flag = 0
        j = i
        current = ''
        index = 0
        while i < len(s):
            
            #current = ''
            #j = i
            print("i = {} j = {}".format(i,j))
            while j < len(s):
                if s[j] not in current:
                    current = current + s[j]
                else:
                    break
                print('i = {} j = {} current = {}'.format(i,j, current))            
                j = j + 1
                

            if len(current) > len(longest):
                longest = current 
            if j < len(s):
                c = s[j]
                index = current.find(c,i)
                current = current[index+1:]
                print('index = {} current = {}'.format(index,current))
                
                i = index + 1
            else:
                i = i + 1
            #    break
            #i = i+1
            if len(longest) + i > len(s):
                break
            flag = flag + 1
            if flag > 10:
                break
            
        return longest

def main():
    a = Solution()
    s = ''
    s = '1234345'
    #s = 'anviaj'
    #s = "abcabcbb"
    #s = 'bbbbb'
    #s = 'pwwkew'
    #s = 'bpfbhmipx'
    ans = a.lengthOfLongestSubstring(s)
    print(ans)


if __name__ == '__main__':
    main()
